contracts/libraries/LibSortValidatorsByBeacon.sol

```diff
@@ -0,0 +1,235 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.9;
+
+import { LibArray } from "./LibArray.sol";
+import { IndexMap, LibIndexMap } from "./LibIndexMap.sol";
+import { governanceValidatorFilter, standardValidatorFilter, nonStandardValidatorFilter } from "../utils/Filters.sol";
+
+library LibSortValidatorsByBeacon {
+  using LibArray for uint256[];
+  using LibArray for address[];
+
+  /// @dev Error thrown when the length of the input arrays do not match.
+  error ErrLengthMismatch();
+
+  /// @dev value is equal to keccak256(abi.encode(uint256(keccak256("@ronin.RandomBeacon.storage.validatorPerPeriod")) - 1)) & ~bytes32(uint256(0xff))
+  bytes32 private constant ValidatorPerPeriodStorageLocation =
+    0x33872be455b999eae35ecf27cbd92ec78caae46559d60fd7688ea114abdbb200;
+
+  struct SortedValidator {
+    uint16 nRV;
+    address[] nonRotatingValidators;
+    RotatingValidator[] rotatingValidators;
+  }
+
+  struct RotatingValidator {
+    address id;
+    uint96 staked;
+  }
+
+  struct UnsortedValidator {
+    // An array of unsorted validator addresses.
+    address[] ids;
+  }
+
+  struct Validator {
+    // A boolean value indicating whether all the validators are saved.
+    bool savedAll;
+    // Use when rotating validators are changed per epoch
+    SortedValidator sorted;
+    // Use when all the validators are saved (when number of validators is less than the sum of nGV, nSV, and nRV)
+    UnsortedValidator unsorted;
+  }
+
+  /**
+   * @dev Requests the sort validator set based on the given parameters.
+   * @param period The period value.
+   * @param nGV The number of governance validators.
+   * @param nSV The number of standard validators.
+   * @param nRV The number of non-standard validators.
+   * @param cids The array of validator addresses.
+   * @param stakeds The array of validator staked amounts.
+   * @param trustedWeights The array of trusted weights for validators.
+   */
+  function requestSortValidatorSet(
+    uint256 period,
+    uint256 nGV,
+    uint256 nSV,
+    uint256 nRV,
+    address[] memory cids,
+    uint256[] memory stakeds,
+    uint256[] memory trustedWeights
+  ) internal {
+    unchecked {
+      uint256 length = cids.length;
+      if (!(length == stakeds.length && length == trustedWeights.length)) revert ErrLengthMismatch();
+      // if the number of validators is less than the sum of `nGV`, `nSV`, and `nRV`, save all cids
+      if (nGV + nSV + nRV >= length) {
+        _saveUnsorted(period, cids);
+        return;
+      }
+
+      // create indices for validators
+      uint256[] memory indices = LibArray.arange(length);
```

Comment:
More comment on "why this indices are not hidden inside the filter"? If it's for gas saving purpose, write a comment.

Reply:
if any else solution to hide the indices but still gas saving, consider to adopt it